### PR TITLE
widevine-cdm: add aarch64 support

### PR DESCRIPTION
**Pull request by dxwil**
## Description of changes

Adds support for arm64 widevine using code from [widevine-installer](https://github.com/AsahiLinux/widevine-installer). The package builds and outputs the relevant files in the same locations as the x64 version, although I cannot test it because I don't know how to make chromium.enableWideVine select a local version, it tries to install the x64 one (please help me on this one :).

There are also some things I would like input on from someone else:
1. Versioning between the two builds. Currently arm is using an older version, it's probably a good idea to make them the same, but without being able to test, I can't do that, because I don't know if newer versions exist and can be successfully patched for arm. List of chromeos images containing arm64 widevine [here](https://gsdview.appspot.com/chromeos-localmirror/distfiles/?marker=distfiles%2Fchromeos-lacros-amd64-squash-zstd-106.0.5249.42).
2. Firefox support. Shouldn't be hard to implement if we come up with a system. Chromium has the enableWideVine flag that as I'm aware does not exist for Firefox. All that Firefox needs are some options as well as one env variable, refer to [this](https://gist.github.com/psanford/1f70a0838c308d28478cf27a9b55c43d?permalink_comment_id=5165332#gistcomment-5165332). I could build this into the package but that would mean installing the package would configure Firefox automatically, which is different from the behavior that exists with Chromium (where enableWideVine needs to be set).

Edit: after some googling I see that Firefox drm works out of the box for x64, but for arm the changes linked above are still necessary.

Edit 2: What I don't understand is why Firefox doesn't need a seperate flag for widevine like chromium, is it because Firefox only installs widevine if drm is explicitly enabled and chromium auto installs it on launch? Could it also have something to do with licenses, why isn't Firefox unfree when it has widevine and chromium is?

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by RossComputerGuy** [comment]
Cool, thank you for putting in the effort. I was meaning to get to this myself but just haven't found the time. I'll review and test this tonight.

File: pkgs/applications/networking/browsers/misc/widevine-cdm.nix
Comment (dxwil) [inline pkgs/applications/networking/browsers/misc/widevine-cdm.nix:60]: Cannot verify without being able to test

**Comment by RossComputerGuy** [comment]
Result of `nixpkgs-review pr 337261` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>widevine-cdm</li>
  </ul>
</details>

File: pkgs/applications/networking/browsers/misc/widevine-cdm.nix
Comment (RossComputerGuy) [inline pkgs/applications/networking/browsers/misc/widevine-cdm.nix:45]: This could be `$widevineInstaller` once `widevine-installer` is `widevineInstaller`.

File: pkgs/applications/networking/browsers/misc/widevine-cdm.nix
Comment (RossComputerGuy) [inline pkgs/applications/networking/browsers/misc/widevine-cdm.nix:61]: Probably could just use `lib.optionalString`?

File: pkgs/applications/networking/browsers/misc/widevine-cdm.nix
Comment (RossComputerGuy) [inline pkgs/applications/networking/browsers/misc/widevine-cdm.nix:55]:
```suggestion
    install -vD libwidevinecdm.so $out/share/google/chrome/WidevineCdm/_platform_specific/linux_${
      if stdenv.targetPlatform.isAarch64 then "arm64"
      else if stdenv.targetPlatform.isx86_64 then "x64"
      else throw "Unsupported platform ${stdenv.targetPlatform.config}"
    }/libwidevinecdm.so
```

**Review by RossComputerGuy — CHANGES_REQUESTED** [review]
Good start, got some changes.

File: pkgs/applications/networking/browsers/misc/widevine-cdm.nix
Comment (dxwil) [reply pkgs/applications/networking/browsers/misc/widevine-cdm.nix:25]: Not really, but I'm not sure in what other way I could download the python script.

File: pkgs/applications/networking/browsers/misc/widevine-cdm.nix
Comment (dxwil) [reply pkgs/applications/networking/browsers/misc/widevine-cdm.nix:61]: in postInstall, or move it to installPhase?

**Comment by dxwil** [comment]
I messed up something with the commits... now the add maintainers one is called the same as the other ones

File: pkgs/applications/networking/browsers/misc/widevine-cdm.nix
Comment (dxwil) [reply pkgs/applications/networking/browsers/misc/widevine-cdm.nix:31]: from my testing fetchurl doesn't know how to unpack a zip file without specifying an unpackPhase

File: pkgs/applications/networking/browsers/misc/widevine-cdm.nix
Comment (RossComputerGuy) [reply pkgs/applications/networking/browsers/misc/widevine-cdm.nix:25]: You can just use the script directly from the GitHub fetcher.

**Comment by emilylange** [comment]
Unsure if I understand the question right, but to get `chromium` with `enableWideVine = true` from your local nixpkgs checkout, you can use
```shell
NIXPKGS_ALLOW_UNFREE=1 nix-build -E "(import ./. {}).chromium.override { enableWideVine = true; }"
```

**Comment by dxwil** [comment]
> Unsure if I understand the question correctly, but to get `chromium` with `enableWideVine = true` from your local nixpkgs checkout, you can use
> 
> ```shell
> NIXPKGS_ALLOW_UNFREE=1 nix-build -E "(import ./. {}).chromium.override { enableWideVine = true; }"
> ```

 [This](https://github.com/NixOS/nixpkgs/blob/d0e1602ddde669d5beb01aec49d71a51937ed7be/pkgs/applications/networking/browsers/chromium/default.nix#L81C3-L93C1) is the part that's responsible for enabling widevine in chromium, it would fetch `${widevine-cdm}` from upstream. How can I make it use a locally built version because upstream does not support aarch64, and would that also require for me to rebuild chromium?

**Comment by RossComputerGuy** [comment]
> it would fetch `${widevine-cdm}` from upstream.

Where does it do that?

> How can I make it use a locally built version of widevine because upstream does not support aarch64

You can run that command in your branch.

> and would that also require for me to rebuild chromium?

Any dependency or derivation changes to Chromium would cause a rebuild.

**Comment by RossComputerGuy** [comment]
That command works and took very little time on my own hardware (note the time nom-build outputted was because I had built it before).
image

**Comment by RossComputerGuy** [comment]
I did a quick look at Tidal HiFi and we'll have to build from source instead of the prebuilt x86_64 binary. However, I did find this issue when looking into their issues. https://github.com/castlabs/electron-releases/issues/23#issuecomment-1878166709

> 2. The CDM contains video decoders, and thus requires MPEG-LA licensing. As long as the CDM is hosted by Google and dynamically installed from their servers, the license they have is enough. Hosting and/or bundling the CDM outside of Google could incur license requirements, and thus have legal consequences.

I am not 100% certain if this could cause us any legal issues. Just bringing this up so more knowledable individuals could say whether this is a problem or not.

**Comment by dxwil** [comment]
> I did a quick look at Tidal HiFi and we'll have to build from source instead of the prebuilt x86_64 binary. 

Could u elaborate?

**Comment by dxwil** [comment]
> I am not 100% certain if this could cause us any legal issues. Just bringing this up so more knowledable individuals could say whether this is a problem or not.

Comparing to the x64 version, we're only applying one patch, that's the whole difference. Is the patch something that could bring trouble?

**Comment by RossComputerGuy** [comment]
> Could u elaborate?

Can't easily run x64 binaries on arm64

> Comparing to the x64 version, we're only applying one patch, that's the whole difference.

Well, it's to a proprietary executable. That executable contains video decoders which are proprietary as well. As long as the binary isn't redistributed there doesn't seem to be any legal concerns.

**Comment by dxwil** [comment]
I tried tidal-hifi and it obviously doesn't launch because it's only downloading the x64 binaries, I don't know why the nix package shows support for aarch64, it shouldn't in the first place.

Would you say the chrome install as done now is figured out and good enough to merge? I would still like to come up with something for Firefox as well as proper versioning though.

**Comment by RossComputerGuy** [comment]
> Would you say the chrome install as done now is figured out and good enough to merge?

Yeah, Chrome is good enough.

> I would still like to come up with something for Firefox as well as proper versioning though.

I think we should bring the Firefox maintainers into the discussion. Maybe after merge, we create an issue and ping them so it doesn't clutter up this PR.

**Comment by dxwil** [comment]
> > Would you say the chrome install as done now is figured out and good enough to merge?
> 
> Yeah, Chrome is good enough.
> 
> > I would still like to come up with something for Firefox as well as proper versioning though.
> 
> I think we should bring the Firefox maintainers into the discussion. Maybe after merge, we create an issue and ping them so it doesn't clutter up this PR.

Yeah good idea, so anything else from me or can u merge?

**Review by RossComputerGuy — APPROVED** [review]
LGTM

**Comment by dxwil** [comment]
PS. We cannot build tidal hifi from source because it uses Castlab's electron with widevine built-in which does not support aarch64. So I'll just open a PR to remove advertised aarch64 support.

**Comment by dxwil** [comment]
> That command works and took very little time on my own hardware (note the time nom-build outputted was because I had built it before). image

Out of interest, what CPU are you using. I'm running the same command right now and it's been building for an hour and from the task number it looks like it's gonna take another 3 minimum. I'm on an Apple m1.

File: pkgs/applications/networking/browsers/misc/widevine-cdm.nix
Comment (winterqt) [inline pkgs/applications/networking/browsers/misc/widevine-cdm.nix:35]: Why did you copy this from upstream? Just copy the file...?

Comment (dxwil) [reply pkgs/applications/networking/browsers/misc/widevine-cdm.nix:35]: Could you elaborate? unsquashfs wouldn't work without renaming using curl like widevine-installer does, but I only tried a few times, maybe it makes sense to figure this part out and remove curl from the buildInputs if that's what you mean.

Comment (winterqt) [reply pkgs/applications/networking/browsers/misc/widevine-cdm.nix:35]: curl in this case is a glorified copy or symlink. It makes sense to use it upstream because they're downloading from a URL. Please try `cp` or `ln`ing the file to see if that works, it should.

**Comment by K900** [comment]
I see code copy-pasted without understanding from Asahi, I see comments marked as resolved without being addressed, this is not the kind of change we want to have. Reverting in #338262, please open a new PR and iterate.

**Pull request merged**
